// appium-android-driver/build/lib/commands/lua/find.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumBaseDriver = require('appium-base-driver');

var commands = {};

/**
 *
 * 查找lua元素  2017/4/24
 *
 */
commands.findLuaElement = function callee$0$0(strategy, selector) {
    var result;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.findLuaElOrEls(strategy, selector, false));

            case 2:
                result = context$1$0.sent;

                result.IS_LUA = true;
                return context$1$0.abrupt('return', result);

            case 5:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
};

commands.findLuaElements = function callee$0$0(strategy, selector) {
    var result;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.findLuaElOrEls(strategy, selector, true));

            case 2:
                result = context$1$0.sent;

                result.forEach(function (element) {
                    element.IS_LUA = true;
                });
                return context$1$0.abrupt('return', result);

            case 5:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
};

commands.doFindLuaElementOrEls = function callee$0$0(params) {
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(this.testbundle.sendAction('find', params));

            case 2:
                return context$1$0.abrupt('return', context$1$0.sent);

            case 3:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
};

commands.findLuaElOrEls = function callee$0$0(strategy, selector, mult) {
    var context = arguments.length <= 3 || arguments[3] === undefined ? '' : arguments[3];
    var params, element, doFind;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        var _this = this;

        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                this.validateLocatorStrategy(strategy);

                if (selector) {
                    context$1$0.next = 3;
                    break;
                }

                throw new Error("Must provide a selector when finding elements");

            case 3:
                params = {
                    strategy: strategy,
                    selector: selector,
                    context: context,
                    multiple: mult
                };
                element = undefined;

                doFind = function doFind() {
                    return _regeneratorRuntime.async(function doFind$(context$2$0) {
                        while (1) switch (context$2$0.prev = context$2$0.next) {
                            case 0:
                                context$2$0.prev = 0;
                                context$2$0.next = 3;
                                return _regeneratorRuntime.awrap(this.doFindLuaElementOrEls(params));

                            case 3:
                                element = context$2$0.sent;
                                context$2$0.next = 11;
                                break;

                            case 6:
                                context$2$0.prev = 6;
                                context$2$0.t0 = context$2$0['catch'](0);

                                if (!(context$2$0.t0.message && context$2$0.t0.message.match(/An element could not be located/))) {
                                    context$2$0.next = 10;
                                    break;
                                }

                                return context$2$0.abrupt('return', false);

                            case 10:
                                throw context$2$0.t0;

                            case 11:
                                if (!params.multiple) {
                                    context$2$0.next = 15;
                                    break;
                                }

                                return context$2$0.abrupt('return', element && element.length !== 0);

                            case 15:
                                return context$2$0.abrupt('return', !_lodash2['default'].isNull(element));

                            case 16:
                            case 'end':
                                return context$2$0.stop();
                        }
                    }, null, _this, [[0, 6]]);
                };

                context$1$0.prev = 6;
                context$1$0.next = 9;
                return _regeneratorRuntime.awrap(this.implicitWaitForCondition(doFind));

            case 9:
                context$1$0.next = 18;
                break;

            case 11:
                context$1$0.prev = 11;
                context$1$0.t0 = context$1$0['catch'](6);

                if (!(context$1$0.t0.message && context$1$0.t0.message.match(/Condition unmet/))) {
                    context$1$0.next = 17;
                    break;
                }

                // only get here if we are looking for multiple elements
                // condition was not met setting res to empty array
                element = [];
                context$1$0.next = 18;
                break;

            case 17:
                throw context$1$0.t0;

            case 18:
                if (!mult) {
                    context$1$0.next = 22;
                    break;
                }

                return context$1$0.abrupt('return', element);

            case 22:
                if (!(!element || _lodash2['default'].size(element) === 0)) {
                    context$1$0.next = 24;
                    break;
                }

                throw new _appiumBaseDriver.errors.NoSuchElementError();

            case 24:
                return context$1$0.abrupt('return', element);

            case 25:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this, [[6, 11]]);
};

exports['default'] = commands;
module.exports = exports['default'];

// we are fine with this, just indicate a retry

// we want to return false if we want to potentially try again
